File: solutions/csharp/prime-factors/2/PrimeFactors.cs
using System.Collections.Generic;
using System.Linq;

public static class PrimeFactors
{
    public static long[] Factors(long number) =>
        FactorsFor(number)
        .ToArray();

    public static IEnumerable<long> FactorsFor(long i)
    {
        int factor = 2;
        while (i > 1)
        {
            while (i % factor != 0) {
                factor++;
            }

            i /= factor;
            yield return factor;
        }
    }
}

File: solutions/scala/prime-factors/1/src/main/scala/PrimeFactors.scala
import scala.annotation.tailrec

object PrimeFactors {
  @tailrec
  def factors(number: Long, currentFactor: Long = 2, accumulator: List[Long] = Nil): List[Long] = {
    if (number <= 1) {
      accumulator
    }
    else {
      val nextFactor     = findNextFactor(currentFactor, number)
      val notYetFactored = number / nextFactor
      factors(notYetFactored, nextFactor, accumulator :+ nextFactor)
    }
  }


  @tailrec
  private def findNextFactor(factor: Long, notYetFactored: Long): Long =
    if (notYetFactored % factor == 0) {
      factor
    }
    else {
      findNextFactor(factor + 1, notYetFactored)
    }
}
